programs/address-lookup-table/src/processor.rs

```diff
@@ -0,0 +1,388 @@
+            return Err(InstructionError::MissingRequiredSignature);
+        };
+
+        let lookup_table_account_ref = lookup_table_account.try_account_ref()?;
+        let lookup_table_data = lookup_table_account_ref.data();
+        let mut lookup_table = AddressLookupTable::deserialize(lookup_table_data)?;
+
+        if lookup_table.meta.authority.is_none() {
+            return Err(InstructionError::Immutable);
+        }
+        if lookup_table.meta.authority != Some(*authority_account.unsigned_key()) {
+            return Err(InstructionError::IncorrectAuthority);
+        }
+        if lookup_table.addresses.len() >= LOOKUP_TABLE_MAX_ADDRESSES {
+            ic_msg!(
+                invoke_context,
+                "Lookup table is full and cannot contain more addresses"
+            );
+            return Err(InstructionError::InvalidArgument);
+        }
+
+        if new_addresses.is_empty() {
+            ic_msg!(invoke_context, "Must extend with at least one address");
+            return Err(InstructionError::InvalidInstructionData);
+        }
+
+        let new_table_addresses_len = lookup_table
+            .addresses
+            .len()
+            .saturating_add(new_addresses.len());
+        if new_table_addresses_len > LOOKUP_TABLE_MAX_ADDRESSES {
+            ic_msg!(
+                invoke_context,
+                "Extended lookup table length {} would exceed max capacity of {}",
+                new_table_addresses_len,
+                LOOKUP_TABLE_MAX_ADDRESSES
+            );
+            return Err(InstructionError::InvalidInstructionData);
+        }
+
+        let clock: Clock = invoke_context.get_sysvar(&clock::id())?;
+        if clock.slot != lookup_table.meta.last_extended_slot {
+            lookup_table.meta.last_extended_slot = clock.slot;
+            lookup_table.meta.last_extended_slot_start_index =
+                u8::try_from(lookup_table.addresses.len()).map_err(|_| {
+                    // This is impossible as long as the length of new_addresses
+                    // is non-zero and LOOKUP_TABLE_MAX_ADDRESSES == u8::MAX + 1.
+                    InstructionError::InvalidAccountData
+                })?;
+        }
+
+        let lookup_table_meta = lookup_table.meta;
+        drop(lookup_table_account_ref);
+
+        let new_table_data_len = checked_add(
+            LOOKUP_TABLE_META_SIZE,
+            new_table_addresses_len.saturating_mul(PUBKEY_BYTES),
+        )?;
+
+        {
+            let mut lookup_table_account_ref_mut = lookup_table_account.try_account_ref_mut()?;
+            AddressLookupTable::overwrite_meta_data(
+                lookup_table_account_ref_mut.data_as_mut_slice(),
+                lookup_table_meta,
+            )?;
+
+            let table_data = lookup_table_account_ref_mut.data_mut();
+            for new_address in new_addresses {
+                table_data.extend_from_slice(new_address.as_ref());
+            }
+        }
```

Comment:
This block changed in the latest commit. Rather than allocating a new account data vector, this tries to modify the `AccountSharedData` `data` vector directly if it isn't already referenced somewhere.